### DISCUSSION TITLE
How to dynamically set dir attribute based on locale in NextJS 10?

### DISCUSSION THREAD

**Discussion by oahmaro** [vercel/next.js#19049] [Help] [answered]
I am experimenting with the new NextJS internationalisation support, and I would like to add a dir attribute to the html tag and dynamically change it based on active locale, to properly support RTL languages.

An example code would be appreciated.

Thanks

**Comment by puntope** [comment, marked as answer]
In case you want to change HTML structure you need custom document.

Create a file like -> pages/_document.js
```
import Document, { Html, Head, Main, NextScript } from 'next/document'

class MyDocument extends Document {
    static async getInitialProps(ctx) {
        const initialProps = await Document.getInitialProps(ctx)
        // locale is in ctx.locale

        return { ...initialProps, locale: ctx?.locale || 'es' }
    }
   
    render = () => (
        <Html dir={this.props.locale === 'ar' ? 'rtl': 'ltr'} lang={this.props.locale}>
            <Head />
            <body>
                <Main />
                <NextScript />
            </body>
        </Html>
    )
}

export default MyDocument
```

For doing it inside a component. Try this. Using useRouter

TestLocale.jsx
```
import { useRouter } from 'next/router'

const TestLocale = () =>  {
  const {locale} = useRouter()

 // locale provides you with the locale, and you can conditionally do stuff

  return (
    <div dir={locale === 'ar' ? 'rtl' : 'ltr'}>
      Locale: {locale}
    </div>
  )
}

export default TestLocale
```
More info: 

https://nextjs.org/docs/advanced-features/i18n-routing#accessing-the-locale-information
https://nextjs.org/docs/advanced-features/custom-document

**Reply by oahmaro** [reply]
Thanks alot, this is exactly what i am looking for

**Reply by puntope** [reply]
Happy to help

**Reply by aseelban** [reply]
> dir={this.props.locale === 'ar' ? 'rtl': 'ltr'}

Hi, @puntope,
I tried your solution didn't work well for me because { locale: ctx?.locale || 'es' } always return 'es' which is static value, not the current locale route value!
so after stringify the objects from this.props I got the following :
`dir="{
  "__NEXT_DATA__": {
    "props": {
      "pageProps": {}
    },
    "page": "/",
    "query": {},
    "buildId": "development",
    "nextExport": true,
    "autoExport": true,
    "isFallback": false,
    "locale": "ar",
    "locales": [
      "en",
      "fr",
      "es",
      "ar"
    ],
    "defaultLocale": "en"
  },
  "buildManifest": {
    "polyfillFiles": [
      "static/chunks/polyfills.js"
    ],
    "devFiles": [
      "static/chunks/react-refresh.js"
    ],
    "ampDevFiles": [
      "static/chunks/webpack.js",
      "static/chunks/amp.js"
    ],
    "lowPriorityFiles": [
      "static/development/_buildManifest.js",
      "static/development/_ssgManifest.js"
    ],
    "pages": {
      "/": [
        "static/chunks/main.js",
        "static/chunks/webpack.js",
        "static/chunks/pages/index.js"
      ],
      "/_app": [
        "static/chunks/main.js",
        "static/chunks/webpack.js",
        "static/chunks/pages/_app.js"
      ],
      "/_error": [
        "static/chunks/main.js",
        "static/chunks/webpack.js",
        "static/chunks/pages/_error.js"
      ]
    },
    "ampFirstPages": []
  },
  "docComponentsRendered": {},
  "dangerousAsPath": "/",
  "canonicalBase": "/ar",
  "ampPath": "",
  "inAmpMode": false,
  "isDevelopment": true,
  "hybridAmp": false,
  "dynamicImports": [],
  "assetPrefix": "",
  "headTags": [],
  "devOnlyCacheBusterQueryString": "?ts=1606483821362",
  "html": "<div class=\"LangSwitcher_langSwitcher__36ay1\"><div class=\"Navbar_navbar__xfM49\"><div><a href=\"/ar\">Home</a></div><div><a href=\"/ar/about\">About</a></div><div><a href=\"/ar/contact\">Contact</a></div></div><div><a href=\"/fr\">FR</a></div><div><a href=\"/ar\">AR</a></div><div><a href=\"/es\">ES</a></div><div><a href=\"/\">EN</a></div><div></div></div><div class=\"ar_container\"><h1>مرحبا بكم!</h1><p>رسالة ترحيبية تليق بكم</p><br/><p>Current locale: ar</p><p>Default locale: en</p><p>Configured locales: [&quot;en&quot;,&quot;fr&quot;,&quot;es&quot;,&quot;ar&quot;]</p></div>",
  "head": [
    {
      "type": "meta",
      "key": "0",
      "ref": null,
      "props": {
        "charSet": "utf-8"
      },
      "_owner": null,
      "_store": {}
    },
    {
      "type": "title",
      "key": ".0",
      "ref": null,
      "props": {
        "children": "الصفحة الرئيسية"
      },
      "_owner": null,
      "_store": {}
    },
    {
      "type": "meta",
      "key": ".1",
      "ref": null,
      "props": {
        "name": "viewport",
        "content": "initial-scale=1.0, width=device-width"
      },
      "_owner": null,
      "_store": {}
    }
  ],
  "styles": []
}"`
so I got acces to current value of route by adding __NEXT_DATA__ : `dir={this.props.__NEXT_DATA__.locale === 'ar' ? 'rtl': 'ltr'}`
but the dir attribute not change until I do hard refresh.
how do fix it !

**Reply by ibrahimsoliman97** [reply]
My document has the language and direction correctly, but I have to hard refresh the page to change the direction, any solution to do it without a hard refresh?

**Reply by alimehasin** [reply]
Same problem, I have to make a refresh, or use `<a>` element and construct the url manually (which is a bad idea) but it will make the refresh

**Reply by mazenalhabeby** [reply]
to solve this problem of refreshing you should use useRouter in _app file after you created _document file .

 import {useRouter} from 'next/router';

 function MyApp({ Component, pageProps }) {

  const { locale } = useRouter();
  const dir = locale === 'ar' ? 'rtl' : 'ltr';

  useEffect(() => {
    document.documentElement.dir = dir;
  },[dir]);


  return (
        <Component {...pageProps} />
  )
}

**Reply by monafHorany** [reply]
Document's getInitialProps function is not called during client-side transitions, nor when a page is statically optimized.

**Reply by monafHorany** [reply]
> to solve this problem of refreshing you should use useRouter in _app file after you created _document file .
> 
> import {useRouter} from 'next/router';
> 
> function MyApp({ Component, pageProps }) {
> 
> const { locale } = useRouter();
> const dir = locale === 'ar' ? 'rtl' : 'ltr';
> 
> useEffect(() => {
> document.documentElement.dir = dir;
> },[dir]);
> 
> return (
> <Component {...pageProps} />
> )
> }

you saved my life man

**Reply by dmytromod** [reply]
_document file rendered on the server, so react hooks and client interactions as `onClick` will not work on it.

You need to split the logic and set default values (initial values as on hard refresh) in `_document` file:
```
render() {
    const {locale} = this.props.__NEXT_DATA__;
    const dir = locale === 'ar' ? 'rtl' : 'ltr';

    return (
      <Html lang={locale}>
       ..............
     <body dir={dir}>
          <Main />
          <NextScript />
        </body>
      </Html>
    );
```

and add client code on `_app` file or component where you make a lang/dir changes:

```
const router = useRouter();
const {defaultLocale, locale} = router;


React.useEffect(() => {
    const dir = locale === 'he' ? 'rtl' : 'ltr';
    document.querySelector("body").setAttribute("dir", dir);
  }, [locale]);
```

**Reply by AbdElRahman111192** [reply]
@monafHorany  @dmytromod  , thanks a lot

**Reply by foadtkf** [reply]
Bro.. this ans just saved my day🔥

**Comment by usameje** [comment]
I tried the snippet above and the dynamically (lang) attribute disappeared.

Here is the updated snippet:

```jsx
// pages/_document.js

import Document, { Html, Head, Main, NextScript } from 'next/document';

class MyDocument extends Document {
  static async getInitialProps(ctx) {
    const initialProps = await Document.getInitialProps(ctx);
    return { ...initialProps };
  }

  render() {
    const { locale } = this.props.__NEXT_DATA__;
    const dir = locale === 'ar' ? 'rtl' : 'ltr';
    return (
      <Html lang={locale} dir={dir}>
        <Head />
        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    );
  }
}

export default MyDocument;


```

**Comment by HossamMarey** [comment]
to add dynamic changer use 

```
useEffect(() => {
    let dir =locale == "ar" ? "rtl" : "ltr";
    let lang = locale == "ar" ? "ar" : "en";
    document.querySelector("html").setAttribute("dir", dir);
    document.querySelector("html").setAttribute("lang", lang);
  }, [locale]);
```

**Comment by Mood-al** [comment]
hey guys is there a possible way to add conditional css import my idea is to load diffrent css files based on the dir attribute 
i mean when the language is arabic dir="rtl" i want to import main.rtl.css and when it's something else i want to load main.css

so i there any possible way to do this mechanic in next js ??

currently i'm acheiving this using Head component by serving diffrent link tags when the locale is ar 
by doing this `<Head>
        <link rel="stylesheet" href="/styles/custom-bootstrap.css" />
        <link rel="stylesheet" href="/styles/theme.css" />

        {locale == "ar" && (
          <link rel="stylesheet" href="/styles/main.rtl.css" />
        )}
      </Head> `

but i dont know if this is good for performance or not

**Reply by markus-azer** [reply]
Hi @Mood-al 
did you find best practice for this issue?

**Reply by Mood-al** [reply]
Currently I'm adding a rtl class to body tag 
And using it inside main.rtl.css like so 

.rtl h3 {text-align : right } for instance 

And I'm importing the main.rtl.css file right under the main.css so it can override its styles 

I dont know if it's best practice or not but so far i havent had any issues with it 

Or you can do it as i did at the beginning 
I loaded it inside <Head /> component by using link tag

**Reply by m5m5h** [reply]
can i do this with tailwindcss ?

**Comment by SSylvain1989** [comment]
Dynamic html and script work fine for me 

```javascript
import Document, {
  Html, Head, Main, NextScript,
} from 'next/document';

class MyDocument extends Document {
  static async getInitialProps(ctx) {
    const initialProps = await Document.getInitialProps(ctx);
    // locale is in ctx.locale

    return { ...initialProps, locale: ctx?.locale || 'en' };
  }

  render = () => (
    <Html dir={this.props.locale === 'fr' ? 'fr' : 'en'} lang={this.props.locale}>
      <Head />
      <body>
        <Main />
        <NextScript />
        {this.props.locale === 'fr'
          ? (
            <>
            <script dangerouslySetInnerHTML={{
              __html: 'french script',
            }} />
            </>
          ) : (
            <script dangerouslySetInnerHTML={{
              __html: 'english script',
            }} />
          )}
      </body>
    </Html>
  )
}

export default MyDocument;
```

**Comment by OsamaNaxxa** [comment]
Using a functional Component you can just add following code, and for versions above 10 lang attribute is automatically handled.

```
import { Head, Html, Main, NextScript } from "next/document";

function Document(props: any) {

  return (
    <Html dir={props.__NEXT_DATA__.locale === "en" ? "ltr" : "rtl"}>
      <Head>

      </Head>
      <body>
        <Main />
        <NextScript />
      </body>
    </Html>
  );
}

export default Document;
```

**Reply by wedhimd** [reply]
Thanks, It worked for me.

**Comment by aynzad** [comment]
Also, there is a CSS hack to change the direction that may seems very basic but it's a life saver

```
html[lang^='fa'],
html[lang^='ar'],
html[lang^='he'] {
  direction: rtl;
}
```
